### ISSUE TITLE
BUG: test failures on 32-bit x86

### ISSUE THREAD

**Issue by pcmoore** [seccomp/libseccomp#142]
With the current libseccomp master branch I'm seeing a small number of test failures on a 32-bit x86 install of Ubuntu 16.04.6.
```
Regression Test Summary
 tests run: 6943
 tests skipped: 29
 tests passed: 6936
 tests failed: 7
 tests errored: 6
```

**Comment by pcmoore** [comment]
The six errors appear to be from test 46-sim-kill_process:
```
 batch name: 46-sim-kill_process
 test mode:  c
 test type:  bpf-sim
Test 46-sim-kill_process%%001-00001 result:   ERROR 46-sim-kill_process rc=12
Test 46-sim-kill_process%%002-00001 result:   ERROR 46-sim-kill_process rc=12
Test 46-sim-kill_process%%003-00001 result:   ERROR 46-sim-kill_process rc=12
Test 46-sim-kill_process%%004-00001 result:   ERROR 46-sim-kill_process rc=12
Test 46-sim-kill_process%%005-00001 result:   ERROR 46-sim-kill_process rc=12
Test 46-sim-kill_process%%006-00001 result:   ERROR 46-sim-kill_process rc=12
```

**Comment by pcmoore** [comment]
The seven failures appear to be a mix of 06-sim-actions and 48-sim-32b_args:
```
 batch name: 06-sim-actions
 test mode:  c
 test type:  bpf-sim
Test 06-sim-actions%%001-00001 result:   SUCCESS
Test 06-sim-actions%%002-00001 result:   SUCCESS
...
Test 06-sim-actions%%008-00100 result:   FAILURE bpf_sim resulted in KILL_PROCESS
```
```
 batch name: 48-sim-32b_args
 test mode:  c
 test type:  bpf-sim
Test 48-sim-32b_args%%001-00001 result:   SUCCESS
Test 48-sim-32b_args%%002-00001 result:   FAILURE bpf_sim resulted in ALLOW
Test 48-sim-32b_args%%003-00001 result:   SUCCESS
Test 48-sim-32b_args%%004-00001 result:   SUCCESS
Test 48-sim-32b_args%%005-00001 result:   SUCCESS
Test 48-sim-32b_args%%006-00001 result:   FAILURE bpf_sim resulted in ALLOW
Test 48-sim-32b_args%%007-00001 result:   SUCCESS
Test 48-sim-32b_args%%008-00001 result:   FAILURE bpf_sim resulted in ALLOW
Test 48-sim-32b_args%%009-00001 result:   SUCCESS
Test 48-sim-32b_args%%010-00001 result:   SUCCESS
Test 48-sim-32b_args%%011-00001 result:   FAILURE bpf_sim resulted in ALLOW
Test 48-sim-32b_args%%012-00001 result:   SUCCESS
Test 48-sim-32b_args%%013-00001 result:   SUCCESS
Test 48-sim-32b_args%%014-00001 result:   SUCCESS
Test 48-sim-32b_args%%015-00001 result:   FAILURE bpf_sim resulted in ALLOW
Test 48-sim-32b_args%%016-00001 result:   SUCCESS
Test 48-sim-32b_args%%017-00001 result:   FAILURE bpf_sim resulted in ALLOW
Test 48-sim-32b_args%%018-00001 result:   SUCCESS
```

**Comment by pcmoore** [comment]
I suspect the 06-sim-actions test is a result of an old kernel (4.4.0-131-generic), I suspect using the Ubuntu HWE kernel would resolve this, but we should probably try to fix the test first.

**Comment by pcmoore** [comment]
Scratch that last comment regarding an old kernel and test 06-sim-actions, the problem it the test itself and not the kernel.  The libseccomp rule is as follows:
```
seccomp_rule_add(ctx, SCMP_ACT_KILL_PROCESS, SCMP_SYS(stat), 0);
```
... with the stat(2) syscall being number 106 on 32-bit x86, and since this particular test checks syscall 106 we see the failure.

The following patch resolves the problem for 06-sim-actions on 32-bit x86.
```
diff --git a/tests/06-sim-actions.tests b/tests/06-sim-actions.tests
index 40a93ae..0a7bc66 100644
--- a/tests/06-sim-actions.tests
+++ b/tests/06-sim-actions.tests
@@ -15,7 +15,8 @@ test type: bpf-sim
 06-sim-actions all             stat            N               N               N       N       N       N       KILL_PROCESS
 06-sim-actions all             rt_sigreturn    N               N               N       N       N       N       LOG
 06-sim-actions x86             0-2             N               N               N       N       N       N       KILL
-06-sim-actions x86             7-172           N               N               N       N       N       N       KILL
+06-sim-actions x86             7-105           N               N               N       N       N       N       KILL
+06-sim-actions x86             107-172         N               N               N       N       N       N       KILL
 06-sim-actions x86             174-350         N               N               N       N       N       N       KILL
 06-sim-actions x86_64          5-14            N               N               N       N       N       N       KILL
 06-sim-actions x86_64          16-350          N               N               N       N       N       N       KILL
```

**Comment by pcmoore** [comment]
It looks like we need to make sure we force the API level to 3 in test 46-sim-kill_process.
```
diff --git a/tests/46-sim-kill_process.c b/tests/46-sim-kill_process.c
index 7ab0725..961a047 100644
--- a/tests/46-sim-kill_process.c
+++ b/tests/46-sim-kill_process.c
@@ -36,6 +36,10 @@ int main(int argc, char *argv[])
        if (rc < 0)
                goto out;
 
+       rc = seccomp_api_set(3);
+       if (rc != 0)
+               return -rc;
+
        ctx = seccomp_init(SCMP_ACT_KILL_PROCESS);
        if (ctx == NULL)
                return ENOMEM;
```

**Comment by pcmoore** [comment]
It appears as though test 48-sim-32b_args assumes a 64-bit system, the fix is likely to force it to use x86_64 as we do with other tests.

**Comment by pcmoore** [comment]
On closer inspection we can't force test 48 to use x86_64 since we are using invalid syscall numbers, e.g. 1000, and the ABI translation code in libseccomp can't handle the conversion.  Test 48 should only work on 64-bit systems, but since we don't currently have an easy way to restrict the test to just 64-bit systems let's limit it to x86_64 for now.
```
diff --git a/tests/48-sim-32b_args.tests b/tests/48-sim-32b_args.tests
index b19fe5e..34ae608 100644
--- a/tests/48-sim-32b_args.tests
+++ b/tests/48-sim-32b_args.tests
@@ -8,24 +8,24 @@
 test type: bpf-sim
 
 # Testname             Arch    Syscall Arg0                    Arg1    Arg2    
-48-sim-32b_args                all     1000    0x0                     N       
-48-sim-32b_args                all     1000    0xffffffff              N       
-48-sim-32b_args                all     1000    0xffffffffffffffff      N       
-48-sim-32b_args                all     1032    0x0                     N       
-48-sim-32b_args                all     1032    0xffffffff              N       
-48-sim-32b_args                all     1032    0xffffffffffffffff      N       
-48-sim-32b_args                all     1064    0x0                     N       
-48-sim-32b_args                all     1064    0xffffffff              N       
-48-sim-32b_args                all     1064    0xffffffffffffffff      N       
-48-sim-32b_args                all     2000    0x0                     N       
-48-sim-32b_args                all     2000    0xffffffff              N       
-48-sim-32b_args                all     2000    0xffffffffffffffff      N       
-48-sim-32b_args                all     2032    0x0                     N       
-48-sim-32b_args                all     2032    0xffffffff              N       
-48-sim-32b_args                all     2032    0xffffffffffffffff      N       
-48-sim-32b_args                all     2064    0x0                     N       
-48-sim-32b_args                all     2064    0xffffffff              N       
-48-sim-32b_args                all     2064    0xffffffffffffffff      N       
+48-sim-32b_args                x86_64  1000    0x0                     N       
+48-sim-32b_args                x86_64  1000    0xffffffff              N       
+48-sim-32b_args                x86_64  1000    0xffffffffffffffff      N       
+48-sim-32b_args                x86_64  1032    0x0                     N       
+48-sim-32b_args                x86_64  1032    0xffffffff              N       
+48-sim-32b_args                x86_64  1032    0xffffffffffffffff      N       
+48-sim-32b_args                x86_64  1064    0x0                     N       
+48-sim-32b_args                x86_64  1064    0xffffffff              N       
+48-sim-32b_args                x86_64  1064    0xffffffffffffffff      N       
+48-sim-32b_args                x86_64  2000    0x0                     N       
+48-sim-32b_args                x86_64  2000    0xffffffff              N       
+48-sim-32b_args                x86_64  2000    0xffffffffffffffff      N       
+48-sim-32b_args                x86_64  2032    0x0                     N       
+48-sim-32b_args                x86_64  2032    0xffffffff              N       
+48-sim-32b_args                x86_64  2032    0xffffffffffffffff      N       
+48-sim-32b_args                x86_64  2064    0x0                     N       
+48-sim-32b_args                x86_64  2064    0xffffffff              N       
+48-sim-32b_args                x86_64  2064    0xffffffffffffffff      N       
 
 test type: bpf-sim-fuzz
 
```

**Comment by pcmoore** [comment]
This should be resolved in 584ca7a5e69d87a4c2c4e4c07ce8415fa59e1351.
